### ISSUE TITLE
Multiple confirms for the same IncrementalDataSet

### ISSUE THREAD

**Issue by PtrBld** [kedro-org/kedro#2457]
## Description
A node may have a confirms attribute with the name of a IncrementalDataSet in the DataCatalog. We have a situations where 2 different nodes depend on a IncrementalDataSet (see image) but would need two different checkpoints since we can not ensure that both of them run at the same time but are rather executed independently.

<img width="684" alt="SCR-20230322-ouu" src="https://user-images.githubusercontent.com/7523956/226981223-6feafa9a-537b-4aa3-8c32-cb5b68617999.png">

We now would want to specify different confirms on each node that would create two different checkpoints in the IncrementalDataSet. 

## Possible Implementation
We could make the confirms function take a list of ConfirmObjects that could contain the name of the dataset and the name of the checkpoint file. We could add the checkpoint file name also in the load_args of the IncrementalDataSet so that the incremental data set knows what data should be included. This way IncrementalDataSets could be used independently by different downstream nodes. Also one node could confirm multiple IncrementalDataSets

What do you think?

**Comment by noklam** [comment]
Is this because two nodes are dependent on the same incremental dataset? I am not sure what's the problem here, did you see different input from two nodes?

**Comment by PtrBld** [comment]
Yes that is the problem. Let´s say I run the "tokenize" node first and a checkpoint is created for the IncrementalDataSet and then I run the "extract" node and in this node no data is loaded since the checkpoint is set to the latest partition.

**Comment by PtrBld** [comment]
This actually works as soon as we set the filepath for the checkpoint in the catalog entry manually (checkpoint.filepath) and remove an assertion in the kedro code. 

https://github.com/kedro-org/kedro/blob/2e70dec396567d2ba38456179aad0d4ae8e83b31/kedro/pipeline/pipeline.py#L849-L857

I wonder why this assertion is there? Could you tell me about the reason?

**Comment by noklam** [comment]
@PtrBld From my understanding `IncrementalDataSet` was created with a specific goal, mainly to handle `Date` type data, imagine use cases like partitioning dataset into Date. That's why the "incremental" need to work with a monotonic filename, it assumes something like `20230102`, `20230103` which clearly the latter is an "increment" of the previous one. Where `apple`, `orange`, and `banana` does not have a specific order.

You can find one of the older tutorial here: https://www.youtube.com/watch?v=v7JSSiYgqpg

To summarize, the problem is that `IncrementalDataSet` is not working when it's an input for multiple nodes?

**Comment by PtrBld** [comment]
Yes you are right. In my case, we are using ''Date'' type partitions. And you are also correct with your summary. When the IncrementalDataSet is input for more than one node and multiple nodes "confirm" the IncrementalDataSet
